r1cs/gadgets/std/src/groups/group_vec.rs

```diff
@@ -0,0 +1,402 @@
+use std::borrow::Borrow;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use algebra::{EndoMulCurve, Field, Group, GroupVec};
+use r1cs_core::{ConstraintSystemAbstract, SynthesisError};
+use crate::boolean::Boolean;
+use crate::eq::EqGadget;
+use crate::groups::{EndoMulCurveGadget, GroupGadget};
+use crate::prelude::{CondSelectGadget, UInt8};
+use crate::{ToBitsGadget, ToBytesGadget};
+use crate::alloc::{AllocGadget, ConstantGadget};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct GroupGadgetVec<ConstraintF: Field, G: Group, GG: GroupGadget<G, ConstraintF>> {
+    vec: Vec<GG>,
+    _field: PhantomData<ConstraintF>,
+    _group: PhantomData<G>,
+}
+
+impl<ConstraintF: Field, G: Group, GG: GroupGadget<G, ConstraintF>> GroupGadgetVec<ConstraintF, G, GG> {
+
+    pub fn new(items: Vec<GG>) -> Self { Self {
+        vec: items,
+        _field: PhantomData,
+        _group: PhantomData,
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            vec: Vec::with_capacity(capacity),
+            _field: PhantomData,
+            _group: PhantomData
+        }
+    }
+
+    pub fn get_vec(&self) -> Vec<GG> { self.vec.clone() }
+
+    pub fn len(&self) -> usize {
+        self.vec.len()
+    }
+
+    pub fn push(&mut self, item: GG) {
+        self.vec.push(item)
+    }
+
+    pub fn iter(&self) -> impl DoubleEndedIterator<Item=&GG> {
```

Comment:
Because we need to iterate over the vector in reverse order when combining commitments with Horner scheme: https://github.com/HorizenOfficial/ginger-lib/blob/add_poly-commit_verifier_gadget/proof-systems/src/darlin/poly-commit/src/domain_extended/constraints/mod.rs#L162